### PR TITLE
Change kube-proxy so that we won't perform no-op calls

### PR DESCRIPTION
**Pull request by wojtek-t**

### CODE REVIEW

**Comment by wojtek-t** [comment]
@k8s-bot bazel test this

**Comment by wojtek-t** [comment]
@thockin - should work now

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -241,17 +250,57 @@ func newServiceChangeMap() serviceChangeMap {
...
+	return true
+}
+
+func (sm *proxyServiceMap) merge(other proxyServiceMap) sets.String {
+	existingPorts := sets.NewString()
+	for serviceName, info := range other {
+		existingPorts.Insert(serviceName.Port)
+		_, exists := (*sm)[serviceName]
+		if !exists {
+			glog.V(1).Infof("Adding new service %q at %s:%d/%s", serviceName, info.clusterIP, info.port, info.protocol)
```
Comment (kargakis) [inline pkg/proxy/iptables/proxier.go]: Do these need to be 1? Seem more like 4. How often is this code  running?

Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: I'm fine with changing those, but it used to be like that in the past (this is not introducing new logs, it just preserves the existing behavior). I will leave the final call to @thockin

**Comment by wojtek-t** [comment]
@thockin - rebased after revert of the PR removing reasons.

**Comment by wojtek-t** [comment]
@k8s-bot etcd3 test this

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -664,30 +673,33 @@ func updateServiceMap(
 
 func (proxier *Proxier) OnEndpointsAdd(endpoints *api.Endpoints) {
 	namespacedName := types.NamespacedName{Namespace: endpoints.Namespace, Name: endpoints.Name}
-	proxier.endpointsChanges.update(&namespacedName, nil, endpoints)
-
-	proxier.syncProxyRules(syncReasonEndpoints)
+	if proxier.endpointsChanges.update(&namespacedName, nil, endpoints) {
+		// TODO: Check if endpoints and services are synced.
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: I don't know what this TODO means?

File: pkg/proxy/iptables/proxier.go [hunk 3]
```diff
@@ -664,30 +673,33 @@ func updateServiceMap(
...
+		proxier.syncProxyRules(syncReasonEndpoints)
+	}
 }
 
 func (proxier *Proxier) OnEndpointsSynced() {
 	proxier.mu.Lock()
 	proxier.endpointsSynced = true
 	proxier.mu.Unlock()
-
+	// Call it unconditionally - this is called once per lifetime.
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: The comment says unconditional, but the reason is not unconditional

File: pkg/proxy/iptables/proxier.go [hunk 4]
```diff
@@ -704,14 +716,10 @@ func updateEndpointsMap(
 		changes.Lock()
 		defer changes.Unlock()
 		for _, change := range changes.items {
-			oldEndpointsMap := endpointsToEndpointsMap(change.previous, hostname)
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: the `hostname` arg is not needed now?

File: pkg/proxy/iptables/proxier.go [hunk 5]
```diff
@@ -538,30 +578,33 @@ func (proxier *Proxier) SyncLoop() {
 
 func (proxier *Proxier) OnServiceAdd(service *api.Service) {
 	namespacedName := types.NamespacedName{Namespace: service.Namespace, Name: service.Name}
-	proxier.serviceChanges.update(&namespacedName, nil, service)
-
-	proxier.syncProxyRules(syncReasonServices)
+	if proxier.serviceChanges.update(&namespacedName, nil, service) {
+		// TODO: Check if endpoints and services are synced.
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: I don't understand this TODO

File: pkg/proxy/iptables/proxier.go [hunk 6]
```diff
@@ -538,30 +578,33 @@ func (proxier *Proxier) SyncLoop() {
...
+		proxier.syncProxyRules(syncReasonServices)
+	}
 }
 
 func (proxier *Proxier) OnServiceSynced() {
 	proxier.mu.Lock()
 	proxier.servicesSynced = true
 	proxier.mu.Unlock()
-
+	// Call it unconditionally - this is called once per lifetime.
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: same comment as endpoints - it says unconditional, but the arg is not.

File: pkg/proxy/iptables/proxier.go (hunk 1 above)
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: leave as-is, we can log-scrub later

File: pkg/proxy/iptables/proxier.go [hunk 7]
```diff
@@ -652,9 +634,9 @@ func updateServiceMap(
 		changes.Lock()
 		defer changes.Unlock()
 		for _, change := range changes.items {
-			mergeSyncRequired, existingPorts := serviceMap.mergeService(change.current)
-			unmergeSyncRequired := serviceMap.unmergeService(change.previous, existingPorts, staleServices)
-			syncRequired = syncRequired || mergeSyncRequired || unmergeSyncRequired
+			existingPorts := serviceMap.merge(change.current)
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: endpoints is simply unmerge old, then merge new.  why is this more complex?  I feel like I figured it out last week and then forgot.  Comment?

File: pkg/proxy/iptables/proxier.go [hunk 8]
```diff
@@ -216,23 +217,31 @@ type serviceChangeMap struct {
...
+		change.previous = endpointsToEndpointsMap(previous, ecm.hostname)
 		ecm.items[*namespacedName] = change
 	}
-	change.current = current
+	change.current = endpointsToEndpointsMap(current, ecm.hostname)
+	if reflect.DeepEqual(change.previous, change.current) {
+		delete(ecm.items, *namespacedName)
+		return false
+	}
+	// TODO: Instead of returning true/false, shouldn't we just always return whether the map
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: This is effectively what it is returning, I think?

File: pkg/proxy/iptables/proxier.go [hunk 9]
```diff
@@ -250,17 +250,57 @@ func newServiceChangeMap() serviceChangeMap {
...
+		change.previous = serviceToServiceMap(previous)
 		scm.items[*namespacedName] = change
 	}
-	change.current = current
+	change.current = serviceToServiceMap(current)
+	if reflect.DeepEqual(change.previous, change.current) {
+		delete(scm.items, *namespacedName)
+		return false
+	}
+	// TODO: Instead of returning true/false, shouldn't we just always return whether the map
```
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: same as endpoints - this is effectively what you are TODO'ing, right?

**Review by thockin — COMMENTED** [review]
First and second are easy.  For some reason I thought the 3rd had to be more complicated.  I can't see anything wrong with it.

Just a few nits.

Once this is in, we can remove the reasons again, right? (along with the short-circuit test in syncProxyRules).  I think we no longer need to track syncRequired.  Am I missing anything?

**Comment by wojtek-t** [comment]
> Once this is in, we can remove the reasons again, right? (along with the short-circuit test in syncProxyRules). I think we no longer need to track syncRequired. Am I missing anything?

Yes - I already have a PR exactly doing this in #46080 (on top of this PR). Once this is merged, I will rebase that one.

Will apply the comments not later than on Monday.

File: pkg/proxy/iptables/proxier.go (hunk 8 above)
Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: What I mean here was to protect against the situation like that:
- update endpoint E1 (this will return true), but then the call is throttled
- updated endpoint E2 (this will return true), but the the call is throttled
- revert endpoint E1 (this will return false)
So the we won't call syncProxyRules, but in fact there is one non-applied change (endpoint E2).

This is probably not very common usecase, so maybe we shouldn't worry that much about it. So just clarified a TODO for now.

File: pkg/proxy/iptables/proxier.go (hunk 9 above)
Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: Same answer as above.

File: pkg/proxy/iptables/proxier.go (hunk 5 above)
Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: Sorry. What I meant here is that if initial syncing of the informer fo either endpoints or services is not finished (this is happening exactly once after kube-proxy restart), then we are not calling syncProxyRules, to avoid situation with updating it with partial state (if it was correct before). 

So what we can loose without solving this TODO is that we can unnecessarily call syncProxyRules when kube-proxy is just starting and loose a tocken. But that only after kube-proxy (re)start so it's not that critical I guess.

For now I clarified the TODO. I can solve it in a separate PR.

File: pkg/proxy/iptables/proxier.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: Same answer as above - clarified the TODO. And will send a separate PR to fix it (it doesn't matter until we don't have periodic runner).

File: pkg/proxy/iptables/proxier.go (hunk 6 above)
Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: Good point. Though with the next PR (which you already LGTMed) it is unconditional. So leaving as is (the comment will be accurate once the second PR is merged).

File: pkg/proxy/iptables/proxier.go (hunk 3 above)
Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: Good point. Though with the next PR (which you already LGTMed) it is unconditional. So leaving as is (the comment will be accurate once the second PR is merged).

File: pkg/proxy/iptables/proxier.go (hunk 7 above)
Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: I wanted to make this change effectively a no-op.
The problem is that for services we are keeping track of open ports and if the port was open before the change and is still open, we don't want to perform any operations for them.
With endpoints, we don't have such situation - we just want to update the state.

The easiest way to do it that I came up with is what I did. But if you have any other suggestions, I'm happy to change it in a followup PR.

**Comment by wojtek-t** [comment]
@thockin - since all your comments were basically about comments (and TODO clarifications), I'm self-applying lgtm label. 
If you will still have some comments, I'm obviously happy to apply them in a followup PR.

**Comment by wojtek-t** [comment]
@k8s-bot kops test this

File: pkg/proxy/iptables/proxier.go [hunk 10]
```diff
@@ -241,17 +253,60 @@ func newServiceChangeMap() serviceChangeMap {
...
 	change, exists := scm.items[*namespacedName]
 	if !exists {
 		change = &serviceChange{}
-		change.previous = previous
+		change.previous = serviceToServiceMap(previous)
 		scm.items[*namespacedName] = change
 	}
-	change.current = current
+	change.current = serviceToServiceMap(current)
+	if reflect.DeepEqual(change.previous, change.current) {
```
Comment (smarterclayton) [inline pkg/proxy/iptables/proxier.go]: I'm surprised this doesn't show up in profiles - reflect.DeepEqual is usually quite expensive in terms of allocations.

Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: This is happening once per endpoint/service watch event.
Given what is happening in syncProxyRules function in terms of both allocations and cpu usage (it's better after recent changes, but it's still a lot), I'm not that surprised.
FYI - 2 weeks ago we used to allocated ~38GB of memory during load test on 100-node clusters - this was significantly more than apiserver was allocating.
